### PR TITLE
add support for ssd1680 ePaper display controller

### PR DESCRIPTION
**Pull request by MrCSharp22**
## Description

- Add SSD1680 ePaper display controller driver.
- Refactor SSD1681 driver to have a common base class with SSD1680

## Motivation and Context

- Bring support for more ePaper drivers (based on what I have on hand).

## How Has This Been Tested?

- Changes brought by this PR do not affect other classes/features in this repo outside of the ePaper folder.
- Tested by visually inspecting the output on the `SSD1680`-based ePaper display.

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [x] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/master/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

### CODE REVIEW

File: devices/ePaper/Drivers/Ssd168x/Ssd1680/Ssd1680.cs [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+            var frameByteIndex = GetFrameBufferIndex(x, y);
+            var pageByteIndex = frameByteIndex - CurrentFrameBufferPageLowerBound;
+
+            // if the specified point falls in the current page, update the buffer
+            if (CurrentFrameBufferPageLowerBound <= frameByteIndex
+                && frameByteIndex < CurrentFrameBufferPageUpperBound)
+            {
+                /*
+                 * Lookup Table for colors on SSD1681
```
Comment (Ellerbach) [inline devices/ePaper/Drivers/Ssd168x/Ssd1680/Ssd1680.cs:186]: I guess this is the 1680 :-)

File: devices/ePaper/Drivers/Ssd168x/Ssd1681/Ssd1681.cs [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+        /// <remarks>
+        /// The SSD1681 comes with 2 RAMs: a Black and White RAM and a Red RAM.
+        /// Writing to the B/W RAM draws B/W pixels on the panel. While writing to the Red RAM draws red pixels on the panel (if the panel supports red).
+        /// However, the SSD1681 doesn't support specifying the color level (no grayscaling), therefore the way the buffer is selected 
+        /// is by performing a simple binary check: 
+        /// if R >= 128 and G == 0 and B == 0 then write a red pixel to the Red Buffer/RAM
+        /// if R == 0 and G == 0 and B == 0 then write a black pixel to B/W Buffer/RAM
+        /// else, assume white pixel and write to B/W Buffer/RAM.
+        /// </remarks>
+        public override void DrawPixel(int x, int y, Color color)
```
Comment (Ellerbach) [inline devices/ePaper/Drivers/Ssd168x/Ssd1681/Ssd1681.cs:174]: for me, this seems the exact same function as in the 1680. What do I miss?
If it's the same, why not put it in the base class?

File: devices/ePaper/Drivers/Ssd168x/Ssd1681/Ssd1681.cs [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+        }
+
+        /// <summary>
+        /// Draws the specified buffer directly to the Black/White RAM on the display.
+        /// Call <see cref="Ssd168x.PerformFullRefresh"/> after to update the display.
+        /// </summary>
+        /// <param name="startXPos">X start position.</param>
+        /// <param name="startYPos">Y start position.</param>
+        /// <param name="buffer">The buffer array to draw.</param>
+        public void DirectDrawBuffer(int startXPos, int startYPos, params byte[] buffer)
```
Comment (Ellerbach) [inline devices/ePaper/Drivers/Ssd168x/Ssd1681/Ssd1681.cs:240]: same for this functions and few others. So I really may have missed something

**Review by Ellerbach — COMMENTED** [review]
Thanks for adding the new driver. I checked the 2 drivers and most of the code is fully the same. So I may have missed something somewhere but seems you can even more refactor to put in the base class.

File: devices/ePaper/Drivers/Ssd168x/Ssd1681/Ssd1681.cs (hunk 3 above)
Comment (MrCSharp22) [reply devices/ePaper/Drivers/Ssd168x/Ssd1681/Ssd1681.cs:240]: Yup, you're right. My original thinking was that there might be other SSD168x controllers with their own initialization sequence and buffer requirements which is why I didn't have those in the base-class. I did a bit more digging on the SSD168x line of controllers and seems like there is the SSD1680(A), SSD1681, and SSD1683 (which I don't have). I skimmed through the SSD1683 datasheet and looks like it uses the same commands and has the same initialization sequence, buffer, etc... . I have moved all the code into the base class now.

**Review by Ellerbach — APPROVED** [review]
Looks good now :-) Thanks!

**Pull request merged**